File: lib/matrix.ex
defmodule Matrix do
  @doc ~S"""
  Multiplying Matrices of the form {{a, b}, {b, c}}
  """
  def mul({{a, b}, {b, c}}, {{x, y}, {y, z}}) do
    middle = a*y + b*z
    {
      {a*x + b*y, middle},
      {middle, b*y + c*z}}
  end

  def power(mat, 1), do: mat
  def power(mat, n) do
    m_2 = power(mat, floor(n/2))
    squared = mul(m_2, m_2)
    case rem(n, 2) do
      0 -> squared
      1 -> mul(mat, squared)
    end
  end
end

File: js/matrix.mjs
function mult([[a, b], [_b, c]], [[x, y], [_y, z]]) {
  const middle = a*y + b*z
  const b_y = b * y
  return [ [a*x + b_y, middle],
    [middle, b_y + c*z]]
}

function pow(matrix, n) {
  if (n===1) {
    return matrix
  } else {
    const n_2 = Math.floor(n/2)
    const m_2 = pow(matrix, n_2)
    const squared = mult(m_2, m_2)
    if (n_2 + n_2 === n) {
      return squared
    } else {
      return mult(matrix, squared)
    }
  }
}
export {pow}
